### PR TITLE
Update RegexParser to match latest corefx regex parsing behavior.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/41291

### CODE REVIEW

File: src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs [hunk 1]
```diff
@@ -40,20 +40,19 @@ private SyntaxToken GetStringToken(string text)
             bool runSubTreeTests = true, [CallerMemberName]string name = "",
             bool allowIndexOutOfRange = false,
             bool allowNullReference = false,
-            bool allowOutOfMemory = false)
+            bool allowOutOfMemory = false,
+            bool allowDiagnosticsMismatch = false)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs]: needed as roslyn now tracks the latest corefx parsing.  However, what roslyn actually references in test code is the buggy .net parsing.   So we have to pass a flag saying that it's ok that our behaviors deviate from each other.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs [hunk 2]
```diff
@@ -1505,34 +1483,6 @@ private bool HasProblem(RegexNodeOrToken component)
...
-            // Parsing the right hand side of a - is extremely strange (and most likely buggy) in
-            // the .NET parser. Specifically, the .NET parser will still consider itself on the
-            // right side no matter how many escaped dashes it sees.  So, for example, the following
-            // is legal [a-\-] (even though \- is less than 'a'). Similarly, the following are
-            // *illegal* [b-\-a] and [b-\-\-a].  That's because the range that is checked is
-            // actually "b-a", even though it has all the \- escapes in the middle.
-            //
-            // This is tracked with: https://github.com/dotnet/corefx/issues/31786
-
-            var first = ParseSingleCharacterClassComponent(isFirst: false, afterRangeMinus: true);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs]: this line moved to the single caller.  All teh remainder is what is no longer necessary as corefx no longer has the behavior of continually ignoring a sequence of `\-` constructs after the `-` in the range expression.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs [hunk 3]
```diff
@@ -1219,32 +1219,6 @@ private static RegexOptions OptionFromCode(VirtualChar ch)
 
         private RegexBaseCharacterClassNode ParseCharacterClass()
         {
-            // Note: ScanCharClass is one of the strangest function in the .NET regex parser. Code
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs]: yaaay, and now we get to undo our strangeness to match it.

**Comment by CyrusNajmabadi** [comment]
Tagging @mavasani

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Did the buggy behavior only live in old .NET _core_ versions or is it present in the .NET Framework as well?

**Comment by CyrusNajmabadi** [comment]
> or is it present in the .NET Framework as well?

It's present in .net framework.

That said, it's for an *exceptionallllllllllllllly* corner case.  i only found it myself by fuzzing and doing a really thorough analysis of the .net regex parser.  That's why i filed the original issue at them. I highly doubt any customers have ever hit this given the zero other reports of the issue beyond my own (which was one of many bug reports i found when analyzing their parser).

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski can you rerun the spanish leg for me?

**Comment by jasonmalinowski** [comment]
@jinujoseph can we get a buddy on this to help @CyrusNajmabadi with the Spanish leg? There is a hang of some kind and if there's some flakiness we may need to investigate.

**Comment by dpoeschl** [comment]
Spanish tests passed this time. Will review shortly.

**Comment by dpoeschl** [comment]
@CyrusNajmabadi  Is our general approach going to be to stick with latest, and people on older versions can suppress these cases if it actually impacts them, or turn off diagnostics altogether?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Is our general approach going to be to stick with latest,

Yes.  This was also discussed during the introduction of this feature and was felt to be a reasonable path forward.

> and people on older versions can suppress these cases if it actually impacts them, or turn off diagnostics altogether?

Correct.  Note: so far, any fixes have been in the "so unlikely that no one should ever hit it".  Indeed, that's why .net even feels fine making these *true* breaking changes in the BCL.  

If they were to chance something significant that really could have wide impact, they would likely need to figure out a compat story, and we would probably want to match them.

Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**